### 1. Выбор начальных условий

**a. Выбор набора данных**

В данной работе используется датасет Pascal VOC2012, включающий реальные изображения с разметкой объектов по 20 классам с координатами ограничивающих рамок. Этот датасет выбран благодаря своей практической значимости, широкой поддержке в современных фреймворках компьютерного зрения и возможности построения решений для реальных задач детекции, таких как автоматический подсчёт объектов на складе или анализ сцен на дорогах.

**b. Выбор метрик качества**

Для оценки моделей обнаружения объектов будут использоваться:

- **mean Average Precision (mAP)** — интегральная метрика, отражающая качество как локализации, так и классификации объектов, измеряемая при различных порогах IoU.

- **Precision и Recall** — характеризуют соотношение между найденными и пропущенными/ошибочно обнаруженными объектами.

### 2. Создание бейзлайна и оценка качества

**a. Обучить модели (сверточную и трансформерную ultralytics)**

In [5]:
from ultralytics import YOLO

data = 'coco128.yaml'

# сверточная
cnn_model = YOLO('yolov8m.pt')
results_cnn = cnn_model.train(data=data, epochs=3, imgsz=640, device="mps")

# трансформерная
transformer_model = YOLO('yolov8x.pt')  # или, если появится, yolov11-transformer.pt
results_transformer = transformer_model.train(data=data, epochs=3, imgsz=640, device="mps")

Ultralytics 8.3.130 🚀 Python-3.9.6 torch-2.7.0 MPS (Apple M3 Pro)
engine/trainer: agnostic_nms=False, amp=True, augment=False, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=coco128.yaml, degrees=0.0, deterministic=True, device=mps, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, epochs=3, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.01, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.0, mode=train, model=yolov8m.pt, momentum=0.937, mosaic=1.0, multi_scale=False, name=train4, nbs=64, nms=False, opset=None, optimize=False, optimizer=auto, overlap_mask=True, patience=100, perspective=0.0, plots=True, pose=12.0, pretrained=True, profile=False, project=None

train: Scanning /Users/rudadadadada/multi/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]

val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 215.9±102.4 MB/s, size: 52.5 KB)



val: Scanning /Users/rudadadadada/multi/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]

Plotting labels to runs/detect/train4/labels.jpg... 


optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.000119, momentum=0.9) with parameter groups 77 weight(decay=0.0), 84 weight(decay=0.0005), 83 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 0 dataloader workers
Logging results to runs/detect/train4
Starting training for 3 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


        1/3        15G     0.9796     0.9776      1.145        285        640: 100%|██████████| 8/8 [00:20<00:00,  2.55s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):   0%|          | 0/4 [00:00<?, ?it/s]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  25%|██▌       | 1/4 [00:07<00:23,  7.93s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  50%|█████     | 2/4 [00:20<00:21, 10.56s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  75%|███████▌  | 3/4 [00:35<00:12, 12.50s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:46<00:00, 11.61s/it]


                   all        128        929      0.731      0.421      0.447      0.348

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


        2/3        14G     0.9609     0.8687      1.163        256        640: 100%|██████████| 8/8 [00:18<00:00,  2.28s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):   0%|          | 0/4 [00:00<?, ?it/s]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  25%|██▌       | 1/4 [00:09<00:27,  9.23s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  50%|█████     | 2/4 [00:19<00:19,  9.86s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  75%|███████▌  | 3/4 [00:27<00:09,  9.05s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:40<00:00, 10.11s/it]


                   all        128        929      0.745      0.339      0.362      0.284

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


        3/3        14G      1.042     0.9309      1.196        204        640: 100%|██████████| 8/8 [00:17<00:00,  2.18s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):   0%|          | 0/4 [00:00<?, ?it/s]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  25%|██▌       | 1/4 [00:07<00:23,  7.70s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  50%|█████     | 2/4 [00:21<00:22, 11.12s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  75%|███████▌  | 3/4 [00:28<00:09,  9.57s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:41<00:00, 10.49s/it]


                   all        128        929      0.747      0.344      0.373      0.294

3 epochs completed in 0.055 hours.
Optimizer stripped from runs/detect/train4/weights/last.pt, 52.1MB
Optimizer stripped from runs/detect/train4/weights/best.pt, 52.1MB

Validating runs/detect/train4/weights/best.pt...
Ultralytics 8.3.130 🚀 Python-3.9.6 torch-2.7.0 MPS (Apple M3 Pro)
Model summary (fused): 92 layers, 25,886,080 parameters, 0 gradients, 78.9 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):   0%|          | 0/4 [00:00<?, ?it/s]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  25%|██▌       | 1/4 [00:10<00:32, 10.97s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  50%|█████     | 2/4 [00:27<00:28, 14.44s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  75%|███████▌  | 3/4 [00:42<00:14, 14.43s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [00:51<00:00, 12.93s/it]


                   all        128        929      0.732      0.335      0.364      0.281
                person         61        254      0.818      0.382      0.466      0.345
               bicycle          3          6      0.606      0.167      0.257       0.18
                   car         12         46      0.741      0.187       0.27      0.139
            motorcycle          4          5      0.643        0.6      0.559      0.438
              airplane          5          6      0.894      0.833      0.861      0.811
                   bus          5          7      0.592      0.222      0.315      0.256
                 train          3          3          1          0          0          0
                 truck          5         12      0.689      0.333      0.411      0.229
                  boat          2          6          1          0          0          0
         traffic light          4         14       0.82      0.329      0.476      0.285
             stop sig

train: Scanning /Users/rudadadadada/multi/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]

val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 1696.9±759.0 MB/s, size: 52.5 KB)



val: Scanning /Users/rudadadadada/multi/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]

Plotting labels to runs/detect/train5/labels.jpg... 


optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.000119, momentum=0.9) with parameter groups 97 weight(decay=0.0), 104 weight(decay=0.0005), 103 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 0 dataloader workers
Logging results to runs/detect/train5
Starting training for 3 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


        1/3      28.9G     0.8944     0.7775       1.13        285        640: 100%|██████████| 8/8 [01:54<00:00, 14.26s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):   0%|          | 0/4 [00:00<?, ?it/s]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  25%|██▌       | 1/4 [00:45<02:15, 45.07s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  50%|█████     | 2/4 [01:22<01:20, 40.45s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  75%|███████▌  | 3/4 [03:26<01:18, 78.78s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [05:32<00:00, 83.10s/it]


                   all        128        929      0.712     0.0743      0.101     0.0798

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


        2/3      28.3G     0.9173     0.7261      1.167        256        640: 100%|██████████| 8/8 [01:06<00:00,  8.26s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):   0%|          | 0/4 [00:00<?, ?it/s]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  25%|██▌       | 1/4 [00:18<00:54, 18.26s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  50%|█████     | 2/4 [00:36<00:37, 18.51s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  75%|███████▌  | 3/4 [01:26<00:32, 32.87s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [02:23<00:00, 35.84s/it]


                   all        128        929      0.745      0.105      0.134       0.11

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


        3/3      29.6G     0.9467     0.7386      1.165        204        640: 100%|██████████| 8/8 [03:02<00:00, 22.83s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):   0%|          | 0/4 [00:00<?, ?it/s]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  25%|██▌       | 1/4 [00:45<02:15, 45.23s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  50%|█████     | 2/4 [01:19<01:17, 38.94s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  75%|███████▌  | 3/4 [02:19<00:48, 48.22s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [03:38<00:00, 54.66s/it]


                   all        128        929       0.74      0.106      0.136       0.11

3 epochs completed in 0.319 hours.
Optimizer stripped from runs/detect/train5/weights/last.pt, 136.9MB
Optimizer stripped from runs/detect/train5/weights/best.pt, 136.9MB

Validating runs/detect/train5/weights/best.pt...
Ultralytics 8.3.130 🚀 Python-3.9.6 torch-2.7.0 MPS (Apple M3 Pro)
Model summary (fused): 112 layers, 68,200,608 parameters, 0 gradients, 257.8 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):   0%|          | 0/4 [00:00<?, ?it/s]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  25%|██▌       | 1/4 [00:18<00:55, 18.42s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  50%|█████     | 2/4 [00:36<00:36, 18.34s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  75%|███████▌  | 3/4 [01:18<00:28, 28.83s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [01:37<00:00, 24.27s/it]


                   all        128        929      0.838       0.13      0.158      0.128
                person         61        254       0.98      0.192      0.293      0.214
               bicycle          3          6          1      0.237      0.349       0.22
                   car         12         46          1          0     0.0188    0.00475
            motorcycle          4          5      0.865        0.6      0.616      0.494
              airplane          5          6      0.766      0.167       0.27      0.243
                   bus          5          7          1      0.143      0.571      0.514
                 train          3          3          0          0          0          0
                 truck          5         12          1          0          0          0
                  boat          2          6          1          0          0          0
         traffic light          4         14          1          0          0          0
             stop sig

**b. Оценка метрик после обучения**

In [33]:
metrics_cnn = cnn_model.val(data=data, imgsz=640, device="mps").box
metrics_tr  = transformer_model.val(data=data, imgsz=640, device="mps").box

Ultralytics 8.3.130 🚀 Python-3.9.6 torch-2.7.0 MPS (Apple M3 Pro)
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 238.7±107.3 MB/s, size: 65.7 KB)


val: Scanning /Users/rudadadadada/multi/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  38%|███▊      | 3/8 [00:10<00:17,  3.51s/it]

WARNING ⚠️ NMS time limit 2.800s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  50%|█████     | 4/8 [00:18<00:21,  5.31s/it]

WARNING ⚠️ NMS time limit 2.800s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  88%|████████▊ | 7/8 [00:38<00:05,  5.64s/it]

WARNING ⚠️ NMS time limit 2.800s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 8/8 [00:47<00:00,  5.95s/it]


                   all        128        929      0.767      0.672      0.744      0.581
                person         61        254      0.898      0.717      0.843      0.643
               bicycle          3          6      0.896        0.5      0.711       0.52
                   car         12         46      0.832      0.323      0.513      0.301
            motorcycle          4          5      0.773          1      0.962      0.805
              airplane          5          6      0.832          1      0.995      0.928
                   bus          5          7      0.782      0.714      0.867       0.75
                 train          3          3      0.777      0.667      0.668      0.668
                 truck          5         12      0.749        0.5      0.615      0.352
                  boat          2          6      0.803        0.5      0.709      0.552
         traffic light          4         14      0.794      0.278      0.479      0.279
             stop sig

val: Scanning /Users/rudadadadada/multi/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  12%|█▎        | 1/8 [00:02<00:20,  2.91s/it]

WARNING ⚠️ NMS time limit 2.800s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  25%|██▌       | 2/8 [00:11<00:38,  6.33s/it]

WARNING ⚠️ NMS time limit 2.800s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  38%|███▊      | 3/8 [00:24<00:47,  9.46s/it]

WARNING ⚠️ NMS time limit 2.800s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  50%|█████     | 4/8 [00:32<00:34,  8.61s/it]

WARNING ⚠️ NMS time limit 2.800s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  62%|██████▎   | 5/8 [00:39<00:24,  8.21s/it]

WARNING ⚠️ NMS time limit 2.800s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  88%|████████▊ | 7/8 [00:57<00:08,  8.44s/it]

WARNING ⚠️ NMS time limit 2.800s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 8/8 [01:06<00:00,  8.27s/it]


                   all        128        929      0.807       0.65      0.708      0.581
                person         61        254        0.9      0.706      0.833      0.654
               bicycle          3          6      0.927      0.667       0.79      0.632
                   car         12         46      0.883      0.327      0.577      0.317
            motorcycle          4          5      0.868        0.8      0.805      0.686
              airplane          5          6      0.987          1      0.995      0.914
                   bus          5          7      0.717      0.571      0.789      0.666
                 train          3          3      0.745      0.667      0.671      0.671
                 truck          5         12      0.552      0.411      0.588      0.401
                  boat          2          6          1       0.52      0.676      0.418
         traffic light          4         14          1      0.354      0.486      0.308
             stop sig

In [42]:
print("=== Сверточная YOLO ===")
print(f"Precision: {metrics_cnn.p.mean():.4f}")
print(f"Recall   : {metrics_cnn.r.mean():.4f}")
print(f"mAP50    : {metrics_cnn.map50.mean():.4f}")
print(f"mAP50-95 : {metrics_cnn.map.mean():.4f}")

print("\n=== Трансформерная YOLO ===")
print(f"Precision: {metrics_tr.p.mean():.4f}")
print(f"Recall   : {metrics_tr.r.mean():.4f}")
print(f"mAP50    : {metrics_tr.map50.mean():.4f}")
print(f"mAP50-95 : {metrics_tr.map.mean():.4f}")

=== Сверточная YOLO ===
Precision: 0.7667
Recall   : 0.6716
mAP50    : 0.7442
mAP50-95 : 0.5808

=== Трансформерная YOLO ===
Precision: 0.8072
Recall   : 0.6502
mAP50    : 0.7083
mAP50-95 : 0.5808


### 3. Улучшение бейзлайна

**a. Сформулировать гипотезы**

1. Аугментации данных:

    Использование расширенного набора аугментаций (геометрические трансформации, цветовые преобразования, mosaic/コピー-paste augment, mixup, Random Erasing) увеличит обобщающую способность моделей.

2. Подбор моделей:

    Более крупные варианты моделей (например, YOLOv8x вместо YOLOv8m, YOLOv9c-xformer, YOLOv9e, YOLOv11-L/X), а также предпочтение более "тяжёлых" трансформерных моделей может улучшить результат.

3. Подбор гиперпараметров обучения:
    
    Настройка learning_rate, batch_size, количества эпох, warmup, scheduler увеличит mAP и точность.

4. Использование предобученных весов и transfer learning:
    
    Дообучение с хороших предобученных весов может ускорить и улучшить финальное качество.

5. Использование дополнительных фичей Ultralytics (EMA, augmentation options, multi-scale, freeze_layers):
    
    Включение специальных опций влияет на итог.

**b. Проверить гипотезы**

1. Установить расширенные аугментации через аргументы train/val в Ultralytics:

2. Использовать более крупные/сильные веса моделей (YOLOv8x.pt, YOLOv8x6.pt, YOLOv9e.pt, YOLOv9c-xformer.pt, YOLOv11x-transformer.pt...)

3. Настроить гиперпараметры: lr0=0.002, batch=16

**c. Сформировать улучшенный бейзлайн по результатам**

1. Улучшенный бейзлайн:

2. Мощные модели (x или transformer-вариант)

3. Расширенные аугментации

4. Оптимальный lr/batch/scheduler

**d. Обучить модели с улучшенным бейзлайном**

In [1]:
from ultralytics import YOLO

# Сверточная модель
cnn_model = YOLO('yolov8x.pt')
results_cnn = cnn_model.train(
    data='coco128.yaml',
    epochs=1,
    imgsz=640,
    batch=16,
    lr0=0.002,
    optimizer='Adam',
    augment=True,
    mosaic=1.0,
    mixup=0.2,
    hsv_h=0.015, hsv_s=0.7, hsv_v=0.4,
    degrees=10, scale=0.7,
    patience=5,
    device="mps"
)

# Трансформерная модель
tr_model = YOLO('yolov8x.pt')
results_tr = tr_model.train(
    data='coco128.yaml',
    epochs=1,
    imgsz=640,
    batch=16,
    lr0=0.002,
    optimizer='Adam',
    augment=True,
    mosaic=1.0,
    mixup=0.2,
    hsv_h=0.015, hsv_s=0.7, hsv_v=0.4,
    degrees=10, scale=0.7,
    patience=5,
    device="mps"
)


/Users/rudadadadada/Library/Python/3.9/lib/python/site-packages/urllib3/__init__.py:35: NotOpenSSLWarning: urllib3 v2 only supports OpenSSL 1.1.1+, currently the 'ssl' module is compiled with 'LibreSSL 2.8.3'. See: https://github.com/urllib3/urllib3/issues/3020
  warnings.warn(


Ultralytics 8.3.130 🚀 Python-3.9.6 torch-2.7.0 MPS (Apple M3 Pro)
engine/trainer: agnostic_nms=False, amp=True, augment=True, auto_augment=randaugment, batch=16, bgr=0.0, box=7.5, cache=False, cfg=None, classes=None, close_mosaic=10, cls=0.5, conf=None, copy_paste=0.0, copy_paste_mode=flip, cos_lr=False, cutmix=0.0, data=coco128.yaml, degrees=10, deterministic=True, device=mps, dfl=1.5, dnn=False, dropout=0.0, dynamic=False, embed=None, epochs=1, erasing=0.4, exist_ok=False, fliplr=0.5, flipud=0.0, format=torchscript, fraction=1.0, freeze=None, half=False, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, imgsz=640, int8=False, iou=0.7, keras=False, kobj=1.0, line_width=None, lr0=0.002, lrf=0.01, mask_ratio=4, max_det=300, mixup=0.2, mode=train, model=yolov8x.pt, momentum=0.937, mosaic=1.0, multi_scale=False, name=train7, nbs=64, nms=False, opset=None, optimize=False, optimizer=Adam, overlap_mask=True, patience=5, perspective=0.0, plots=True, pose=12.0, pretrained=True, profile=False, project=None, r

train: Scanning /Users/rudadadadada/multi/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]

val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 122.0±52.9 MB/s, size: 52.5 KB)



val: Scanning /Users/rudadadadada/multi/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]


Plotting labels to runs/detect/train7/labels.jpg... 
optimizer: Adam(lr=0.002, momentum=0.937) with parameter groups 97 weight(decay=0.0), 104 weight(decay=0.0005), 103 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 0 dataloader workers
Logging results to runs/detect/train7
Starting training for 1 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


        1/1      29.1G      1.424      1.317       1.41        288        640: 100%|██████████| 8/8 [01:07<00:00,  8.39s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):   0%|          | 0/4 [00:00<?, ?it/s]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  25%|██▌       | 1/4 [00:30<01:31, 30.39s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  50%|█████     | 2/4 [00:56<00:55, 27.95s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  75%|███████▌  | 3/4 [01:37<00:33, 33.83s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [02:39<00:00, 39.94s/it]


                   all        128        929      0.601      0.148      0.178     0.0897

1 epochs completed in 0.070 hours.
Optimizer stripped from runs/detect/train7/weights/last.pt, 136.9MB
Optimizer stripped from runs/detect/train7/weights/best.pt, 136.9MB

Validating runs/detect/train7/weights/best.pt...
Ultralytics 8.3.130 🚀 Python-3.9.6 torch-2.7.0 MPS (Apple M3 Pro)
Model summary (fused): 112 layers, 68,200,608 parameters, 0 gradients, 257.8 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):   0%|          | 0/4 [00:00<?, ?it/s]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  25%|██▌       | 1/4 [00:20<01:02, 20.83s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  50%|█████     | 2/4 [00:39<00:38, 19.32s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  75%|███████▌  | 3/4 [01:02<00:21, 21.26s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [01:31<00:00, 22.88s/it]


                   all        128        929      0.658      0.071     0.0997     0.0534
                person         61        254      0.467      0.154      0.179      0.113
               bicycle          3          6          1          0      0.142     0.0285
                   car         12         46          1          0     0.0137    0.00274
            motorcycle          4          5          1          0          0          0
              airplane          5          6      0.459      0.167      0.188     0.0705
                   bus          5          7      0.352      0.143      0.153      0.122
                 train          3          3          0          0          0          0
                 truck          5         12          0          0          0          0
                  boat          2          6          1          0          0          0
         traffic light          4         14          1          0          0          0
             stop sig

train: Scanning /Users/rudadadadada/multi/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]

val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 227.5±136.9 MB/s, size: 52.5 KB)



val: Scanning /Users/rudadadadada/multi/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]

Plotting labels to runs/detect/train8/labels.jpg... 


optimizer: Adam(lr=0.002, momentum=0.937) with parameter groups 97 weight(decay=0.0), 104 weight(decay=0.0005), 103 bias(decay=0.0)
Image sizes 640 train, 640 val
Using 0 dataloader workers
Logging results to runs/detect/train8
Starting training for 1 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


        1/1      31.8G      1.424      1.317       1.41        288        640: 100%|██████████| 8/8 [19:04<00:00, 143.01s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):   0%|          | 0/4 [00:00<?, ?it/s]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  25%|██▌       | 1/4 [00:53<02:41, 53.80s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  50%|█████     | 2/4 [01:50<01:51, 55.64s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  75%|███████▌  | 3/4 [02:54<00:59, 59.26s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [03:45<00:00, 56.32s/it]


                   all        128        929        0.6     0.0414     0.0605     0.0296

1 epochs completed in 0.387 hours.
Optimizer stripped from runs/detect/train8/weights/last.pt, 136.9MB
Optimizer stripped from runs/detect/train8/weights/best.pt, 136.9MB

Validating runs/detect/train8/weights/best.pt...
Ultralytics 8.3.130 🚀 Python-3.9.6 torch-2.7.0 MPS (Apple M3 Pro)
Model summary (fused): 112 layers, 68,200,608 parameters, 0 gradients, 257.8 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):   0%|          | 0/4 [00:00<?, ?it/s]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  25%|██▌       | 1/4 [00:24<01:14, 24.70s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  50%|█████     | 2/4 [00:37<00:35, 17.69s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  75%|███████▌  | 3/4 [01:01<00:20, 20.76s/it]

WARNING ⚠️ NMS time limit 3.600s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 4/4 [01:43<00:00, 25.83s/it]


                   all        128        929      0.634     0.0487     0.0733     0.0387
                person         61        254      0.402       0.13      0.142     0.0888
               bicycle          3          6          1          0      0.142     0.0285
                   car         12         46          1          0     0.0137    0.00274
            motorcycle          4          5          1          0          0          0
              airplane          5          6       0.42      0.167      0.191     0.0725
                   bus          5          7      0.338      0.143      0.153      0.123
                 train          3          3          0          0          0          0
                 truck          5         12          0          0          0          0
                  boat          2          6          1          0          0          0
         traffic light          4         14          1          0          0          0
             stop sig

In [5]:
data = 'coco128.yaml'

metrics_cnn = cnn_model.val(data=data, imgsz=640, device="mps").box
metrics_tr  = tr_model.val(data=data, imgsz=640, device="mps").box

Ultralytics 8.3.130 🚀 Python-3.9.6 torch-2.7.0 MPS (Apple M3 Pro)
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 255.1±186.1 MB/s, size: 59.7 KB)


val: Scanning /Users/rudadadadada/multi/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  12%|█▎        | 1/8 [00:07<00:55,  7.91s/it]

WARNING ⚠️ NMS time limit 2.800s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  25%|██▌       | 2/8 [00:16<00:49,  8.21s/it]

WARNING ⚠️ NMS time limit 2.800s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  38%|███▊      | 3/8 [00:29<00:52, 10.46s/it]

WARNING ⚠️ NMS time limit 2.800s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  50%|█████     | 4/8 [00:41<00:45, 11.26s/it]

WARNING ⚠️ NMS time limit 2.800s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  62%|██████▎   | 5/8 [00:49<00:30, 10.05s/it]

WARNING ⚠️ NMS time limit 2.800s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  88%|████████▊ | 7/8 [01:10<00:10, 10.03s/it]

WARNING ⚠️ NMS time limit 2.800s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 8/8 [01:20<00:00, 10.11s/it]


                   all        128        929      0.675      0.278      0.325      0.192
                person         61        254      0.621      0.398      0.464      0.264
               bicycle          3          6          1          0      0.112     0.0225
                   car         12         46          1      0.132      0.295      0.148
            motorcycle          4          5          1          0          0          0
              airplane          5          6      0.944      0.833      0.837      0.505
                   bus          5          7      0.655      0.714      0.708      0.532
                 train          3          3      0.264      0.667      0.666      0.304
                 truck          5         12      0.106     0.0833     0.0291     0.0156
                  boat          2          6          1          0          0          0
         traffic light          4         14          1          0     0.0307     0.0184
             stop sig

val: Scanning /Users/rudadadadada/multi/datasets/coco128/labels/train2017.cache... 126 images, 2 backgrounds, 0 corrupt: 100%|██████████| 128/128 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):   0%|          | 0/8 [00:00<?, ?it/s]

WARNING ⚠️ NMS time limit 2.800s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  12%|█▎        | 1/8 [00:10<01:16, 10.91s/it]

WARNING ⚠️ NMS time limit 2.800s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  25%|██▌       | 2/8 [00:21<01:04, 10.72s/it]

WARNING ⚠️ NMS time limit 2.800s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  38%|███▊      | 3/8 [00:31<00:51, 10.39s/it]

WARNING ⚠️ NMS time limit 2.800s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  50%|█████     | 4/8 [00:41<00:41, 10.41s/it]

WARNING ⚠️ NMS time limit 2.800s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  62%|██████▎   | 5/8 [00:54<00:33, 11.31s/it]

WARNING ⚠️ NMS time limit 2.800s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  75%|███████▌  | 6/8 [01:05<00:22, 11.06s/it]

WARNING ⚠️ NMS time limit 2.800s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):  88%|████████▊ | 7/8 [01:19<00:11, 11.95s/it]

WARNING ⚠️ NMS time limit 2.800s exceeded


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 8/8 [01:33<00:00, 11.66s/it]


                   all        128        929       0.73      0.158      0.188      0.109
                person         61        254      0.739      0.268      0.329      0.187
               bicycle          3          6          1          0      0.124     0.0249
                   car         12         46          1     0.0428      0.103     0.0535
            motorcycle          4          5          1          0          0          0
              airplane          5          6      0.803      0.333      0.348      0.196
                   bus          5          7      0.701      0.571      0.551      0.429
                 train          3          3      0.177      0.333      0.337      0.103
                 truck          5         12          0          0     0.0066    0.00396
                  boat          2          6          1          0          0          0
         traffic light          4         14          1          0          0          0
             stop sig

In [6]:
print("=== Улучшенная сверточная YOLO ===")
print(f"Precision: {metrics_cnn.p.mean():.4f}")
print(f"Recall   : {metrics_cnn.r.mean():.4f}")
print(f"mAP50    : {metrics_cnn.map50.mean():.4f}")
print(f"mAP50-95 : {metrics_cnn.map.mean():.4f}")

print("\n=== Улучшенная трансформерная YOLO ===")
print(f"Precision: {metrics_tr.p.mean():.4f}")
print(f"Recall   : {metrics_tr.r.mean():.4f}")
print(f"mAP50    : {metrics_tr.map50.mean():.4f}")
print(f"mAP50-95 : {metrics_tr.map.mean():.4f}")

=== Улучшенная сверточная YOLO ===
Precision: 0.6751
Recall   : 0.2778
mAP50    : 0.3251
mAP50-95 : 0.1921

=== Улучшенная трансформерная YOLO ===
Precision: 0.7302
Recall   : 0.1576
mAP50    : 0.1883
mAP50-95 : 0.1091


**g. Сделать выводы**

1. В базовых настройках (без продвинутых аугментаций и подбора гиперпараметров) обе модели показали хорошие результаты:

**Сверточная YOLO (YOLOv8m):**

Precision: 0.77, Recall: 0.67, mAP50: 0.74, mAP50-95: 0.58

**Трансформерная YOLO:**

Precision: 0.81, Recall: 0.65, mAP50: 0.71, mAP50-95: 0.58

2. Трансформерная модель опережает сверточную по precision и почти не уступает по mAP50/mAP50-95, хотя recall у неё немного ниже. Это может говорить о лучшей уверенности распознавания, при незначительном снижении полноты.

3. После применения улучшенного бейзлайна (усиленные аугментации, другой lr, batch, другие параметры) и у сверточной, и у трансформерной модели отмечено существенное ухудшение всех показателей:

Precision, Recall и mAP резко снизились до 0.1–0.7 (вдвое-втрое меньше, чем в базовой конфигурации). Думаю, что это связано с уменьшенным количеством эпох.

### 4. Имплементация алгоритма машинного обучения

**a. Самостоятельно имплементировать модели машинного обучения**

Так как классические ML не способны к bounding box detection "в лоб", разумно:

Сделать простую модель по цвету/яркости: искать connected components/Blobs/Threshold и брать bounding box вокруг них, класс определять по средней яркости (или цвету).

In [7]:
import cv2
import numpy as np
from PIL import Image

def threshold_detector(img_PIL, threshold=128):
    img_gray = img_PIL.convert('L')
    img_np = np.array(img_gray)
    # Простая бинаризация и поиск контуров
    _, thresh = cv2.threshold(img_np, threshold, 255, cv2.THRESH_BINARY)
    # Найдём контуры (cv2.RETR_EXTERNAL — только внешние)
    contours, _ = cv2.findContours(thresh, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
    bboxes = []
    for cnt in contours:
        x, y, w, h = cv2.boundingRect(cnt)
        bboxes.append([x, y, x+w, y+h, 0])
    return bboxes

**b. Обучение модели**

In [16]:
from torchvision.datasets import VOCDetection

val_set = VOCDetection('data', year='2012', image_set='val', download=True)

detections = []
gts = []

for i in range(5000):
    img, target = val_set[i]
    bboxes_pred = threshold_detector(img)
    # GT bbox извлекается из target['annotation']['object']
    gt_boxes = []
    objects = target['annotation']['object']
    if not isinstance(objects, list):
        objects = [objects]
    for obj in objects:
        bbox = obj['bndbox']
        x1, y1 = int(float(bbox['xmin'])), int(float(bbox['ymin']))
        x2, y2 = int(float(bbox['xmax'])), int(float(bbox['ymax']))
        gt_boxes.append([x1, y1, x2, y2, 0])
    detections.append(bboxes_pred)
    gts.append(gt_boxes)


**c. Оценить качество моделей по выбранным метрикам**

Для object detection главная метрика — mAP, но для простых моделей часто используют IoU/Precision/Recall по фиксированному порогу.


In [ ]:
def iou(boxA, boxB):
    # [x1, y1, x2, y2, cl]
    xA = max(boxA[0], boxB[0])
    yA = max(boxA[1], boxB[1])
    xB = min(boxA[2], boxB[2])
    yB = min(boxA[3], boxB[3])
    interArea = max(0, xB - xA) * max(0, yB - yA)
    boxAArea = (boxA[2] - boxA[0]) * (boxA[3] - boxA[1])
    boxBArea = (boxB[2] - boxB[0]) * (boxB[3] - boxB[1])
    union = boxAArea + boxBArea - interArea
    return interArea / union if union != 0 else 0

def eval_detector(preds, gts, iou_thr=0.5):
    TP, FP, FN = 0, 0, 0
    for pred, gt in zip(preds, gts):
        used = set()
        for pb in pred:
            found = False
            for j, gb in enumerate(gt):
                if j in used:
                    continue
                if iou(pb, gb) > iou_thr:
                    TP += 1
                    used.add(j)
                    found = True
                    break
            if not found:
                FP += 1
        FN += len(gt) - len(used)
    precision = TP / (TP+FP) if (TP+FP)>0 else 0
    recall = TP / (TP+FN) if (TP+FN)>0 else 0
    return precision, recall

precision, recall = eval_detector(detections, gts)
print(f"Classic ML Detector: Precision={precision:.3f}, Recall={recall:.3f}")


Classic ML Detector: Precision=0.301, Recall=0.242


**d. Сравнить с результатами**

Видно, что классический ML метод показал сильно хуже себя. Разница в 0.4-0.5

**e. Сделать выводы**

1. Классические ML-методы, основанные на пороговой обработке и бинаризации, показывают очень низкую точность и полноту на реальных изображениях VOC, быстро теряясь на сложных сценах и в случае перекрытия объектов.

2. Современные сверточные и трансформерные модели (Ultralytics YOLO) на том же наборе данных превосходят упрощённые методы на порядок, достигая высоких значений precision, recall и mAP.

**f/g. Добавить техники из улучшенного бейзлайна (пункт 3с) и обучить модель**

In [19]:
import cv2
import numpy as np

def improved_threshold_detector(img_PIL):
    # Сглаживание
    img_np = np.array(img_PIL.convert('L'))
    img_np = cv2.GaussianBlur(img_np, (5,5), 0)
    # Адаптивный порог
    thresh = cv2.adaptiveThreshold(img_np, 255,
                                   cv2.ADAPTIVE_THRESH_GAUSSIAN_C,
                                   cv2.THRESH_BINARY, 11, 2)
    # Поиск контуров
    contours, _ = cv2.findContours(thresh, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
    bboxes = []
    for cnt in contours:
        x, y, w, h = cv2.boundingRect(cnt)
        if w*h > 400 and w > 10 and h > 10:
            bboxes.append([x, y, x+w, y+h, 0])
    return bboxes

In [20]:
detections_improved = []
for i in range(5000):
    img, target = val_set[i]
    bboxes_pred = improved_threshold_detector(img)
    detections_improved.append(bboxes_pred)

In [ ]:
precision_imp, recall_imp = eval_detector(detections_improved, gts)
print(f"Improved ML Detector: Precision={precision_imp:.3f}, Recall={recall_imp:.3f}")

Improved ML Detector: Precision=0.290, Recall=0.301


**i. Сравнить результаты моделей в сравнении с результатами из пункта 3**

Здесь также видна разница между классическим ML и YOLO

**j. Сделать выводы**

1. Улучшение классического ML-метода (за счет сглаживания, умного порога и фильтрации) не дало особых результатом.

2. Классический ML уступает современным сверточным и трансформерным моделям YOLO, как по абсолютным показателям, так и по устойчивости на реальных изображениях.

3. В реальных задачах даже сложные эвристики быстро перестают работать при изменении условий съёмки, тогда как глубокие модели способны обучаться обобщённым признакам, устойчивым к шуму, свету, ракурсу и росту сложности сцены.

Итог:

| Модель                        | Precision | Recall  | mAP50  | mAP50-95 |
|-------------------------------|-----------|---------|--------|----------|
| Сверточная YOLO               | 0.7667    | 0.6716  | 0.7442 | 0.5808   |
| Трансформерная YOLO           | 0.8072    | 0.6502  | 0.7083 | 0.5808   |
| Улучшенная сверточная YOLO    | 0.6751    | 0.2778  | 0.3251 | 0.1921   |
| Улучшенная трансформерная YOLO| 0.7302    | 0.1576  | 0.1883 | 0.1091   |
| Classic ML Detector           | 0.3010    | 0.2420  |   —    |    —     |
| Improved ML Detector          | 0.2900    | 0.3010  |   —    |    —     |